### DISCUSSION TITLE
Google Cloud (GCP) Deployment Stack for Langfuse v3

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#4646] [Ideas]
### Describe the feature or potential improvement

Currently it is recommended to use [Helm/Kubernetes](https://langfuse.com/self-hosting/kubernetes-helm) to deploy Langfuse v3. 

Let's collect interest for a GCP specific deployment stack in this thread. Please share what you'd like to see and whether you are interested in contributing to this.

## Edit (2025-04-16)
Today, we released https://github.com/langfuse/langfuse-terraform-gcp which allows you to deploy a production-ready Langfuse stack on GCP using mostly native GCP services. Let us know what you think and share any feedback you have.

**Comment by marcklingen** [comment]
for reference, @sotazum created a terraform repo for GCP

thread: https://github.com/orgs/langfuse/discussions/4664
repo: https://github.com/sotazum/langfuse-terraform

This currently targets Langfuse v2 and could be a starting point for the Langfuse v3 deployment stack for GCP.

**Comment by sotazum** [comment]
I am strongly interested in considering new architecture for Langfuse v3.

 I have an experience architecting Dify on GCP by Terraform using Cloud Run, Memorystore (Redis), Cloud Storage, etc. It will probably be a similar architecture.

reference: https://github.com/DeNA/dify-google-cloud-terraform

**Reply by marcklingen** [reply]
thanks for sharing! would you feel comfortable creating an initial draft based on the helm chart? https://github.com/langfuse/langfuse-k8s

**Reply by sotazum** [reply]
Thanks for good repo. It's helpful to grasp v3 entire architecture. 

I believe there is a demand for serverless hosting with Cloud Run instead of GKE, and I intend to architect in a way that uses Cloud Run!

**Reply by marcklingen** [reply]
makes sense, i think it should be feasible to run this on cloud run (big fan of cloud run myself). i'd suggest to not scale to zero though to make sure that the overall performance is good.

**Reply by MarkEdmondson1234** [reply]
Yes I’m interested in this also and think serverless would be nicer than k8s for my use cases. I’ve also been using pubsub for asynchronous stuff like scores that aren’t needed immediately

**Reply by elliottower** [reply]
Commenting here to get notified about any updates, currently using the langfuse v2 terraform gcp repo linked in the above thread

**Reply by Freezaa9** [reply]
Same here, using as much serverless/managed product on GCP (cloud run, cloud sql, Memorystore , gcs ...) would be the best option for us. It is nearly impossible in a non fully tech company to start to manage and find resources for a k8s cluster for 1 product.

**Reply by gcardoso2314** [reply]
Completely agree with @Freezaa9 - I'm a big fan of the serverless setup and one of the big reasons I chose to self host Langfuse in the first place

**Reply by tkreuder** [reply]
I am also interested in a version without k8s. It would also be awesome, if the new stack of v3 is optional, so fallback to e.g. postgres only is possible.

**Reply by sotazum** [reply]
Hi, I seem to have succeeded in self-hosting langfuse v3 with Cloud Run oriented serverless stack. Let me show the details within a week.

**Comment by adithya-dinesh** [comment]
It would be nice to implement k8 with gke and terraform for it.

**Reply by Steffen911** [reply]
@adithya-dinesh There should be many [quick start guides](https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/using_gke_with_terraform) and templates on GKE and we have https://github.com/langfuse/langfuse-k8s which should be a "one click install" for any running cluster. What would you have in mind as an addition or need specifically for Langfuse?

**Comment by Manouchehri** [comment]
This is a blocker for us. We are currently on (and will likely stay on) Langfuse v2 until it's easier to run v3 on Cloud Run.

**Reply by marcklingen** [reply]
Thanks for sharing and your +1 on this David

**Comment by sotazum** [comment]
Updated and started to support v3 architecture on my terraform repository! Check it out and try to deploy on your project!!
https://github.com/sotazum/langfuse-google-cloud-terraform (renamed langfuse-terraform -> langfuse-google-cloud-terraform)

The architecture and points are below.

- Use Cloud Run for web, worker and clickhouse-server
- Use Filestore for Clickhouse persistent volumes (relatively expensive...)
- Edit Clickhouse migrations to use HTTP because Cloud Run is serverless service and not support TCP connection to the instance.

We may have better way to host Clickhouse such as building GCE cluster using managed instance group. Need to continue to test and discuss this topic.

![langfuse_v3_self_host_google_cloud](https://github.com/user-attachments/assets/c743769c-4ea9-414a-9943-dca95fea4305)

**Reply by MarkEdmondson1234** [reply]
This looks great will give it a go. I use AlloyDB instead of CloudSQL so will adjust that but good to hear Clickhouse is workable as a Cloud Run service

**Comment by Steffen911** [comment]
Hey all,
Today, we released https://github.com/langfuse/langfuse-terraform-gcp which allows you to deploy a production-ready Langfuse stack on GCP using mostly native GCP services. Let us know what you think and share any feedback you have.

**Reply by MarkEdmondson1234** [reply]
Honestly I’d rather keep to my cloud run v2 installation, the services used in the terraform above I wouldn’t call GCP native, but it’s understandable as most migratable to other clouds. Babysitting kubernetes cloud sql etc is not compelling - I’m crossing fingers there is still an option in the future for more what I consider native services - a stack with Redis, Cloud Run, AlloyDB etc. if I get a free week I’ll look at it, I think it was the Clickhouse I was most uncertain about.

**Comment by tc5613213** [comment]
+1 on staying on v2 - a lot of infra overhead to v3, specially for clickhouse. Came across https://github.com/sotazum/langfuse-google-cloud-terraform but still clickhouse requires nfs. If there is a lite version where you simply way to deploy on cloud run would be great.
